### DISCUSSION TITLE
Moving average over turnover per sales item per sell

### DISCUSSION THREAD

**Discussion by poostwoud** [ClickHouse/ClickHouse#65336] [Q&A]
Hi,

Considering the following table:

```sql
create table fact_sales
(
  sellId Int64,
  sellTimestamp DateTime64(8),
  salesItemId Int64,
  turnover Decimal(10, 2)
)
engine = MergeTree
primary key sellId
order by sellId
```

The table has around 1.69 million rows. I'm trying to calculate the movering average over the turnover per sales item using a 1 year interval like this:

```sql
select a.sellId
  ,avg(b.turnover)
from fact_sales a
join fact_sales b
  on b.salesItemId = a.salesItemId
  and b.sellTimestamp between dateadd(year, -1, a.sellTimestamp) and a.sellTimestamp
group by a.sellId
```

Depending on the number of sells per sales item the query becomes really slow. I've tried secondary tables (pk on salesItemId and sellTimestamp), materialized views, projections, but I can't seem to get it right. It is probably the `join` that causes the problem. Given the selection it could be solved using window functions probably. This would eliminate the join, but this would require a time interval to be specified. 

I need to rethink my approach, but I don't know in what direction to look. Any thoughts?

Best regards,

Peter

**Comment by UnamedRus** [comment]
Just a guess

```
SELECT arrayJoin(ids) as sellId, mv_avg FROM (
select  toDate(sellTimestamp ) as date, groupArray(sellId) as ids, avgMerge(avgState(turnover)) OVER (PARTITION BY salesItemId ORDER BY date ASC ROWS  BETWEEN 365 PRECEDING AND CURRENT ROW) as mv_avg
from fact_sales 
group by salesItemId, date)
```

https://kb.altinity.com/altinity-kb-queries-and-syntax/cumulative-unique/

**Reply by poostwoud** [reply]
Hi @UnamedRus,

Thank you for your reply.

Your query has the same issue I keep running into with window functions. It queries a range of 365 rows instead of a 365 days interval. That is why I added the join, to ensure all turnover in the 365 days range (which can be more then 365 rows!) is take into account.

In the documentation I've found that window functions with range interval are not (yet) supported. MySQL seems to support this though, so maybe I can better try to test this way of work on MySQL first.

Best regards,

Peter

**Reply by UnamedRus** [reply]
If you take a deeper look, it first do GROUP BY itemId, date

So in the end there is exactly 1 row per day per item, so it should query exactly 365 days

Another option is to use WINDOW GROUPS clause

**Reply by poostwoud** [reply]
Maybe I'm mistaken, but the group by date eliminates the detail I need. I will try to explain using the following table:

| sellId | sellTimestamp | salesItemId | turnover | ma | comments |
|---|---|---|---|---|---|
| 1 | 2023-01-01 | A | 10 | 10 | Average up to and including sellId 1 |
| 2 | 2023-02-15 | A | 15 | 12,5 | Average up to and including sellId 2 |
| 3 | 2023-02-15 | A | 5 | 10 | Average up to and including sellId 3, etc. |
| 4 | 2023-02-15 | A | 25 | 13,75 |  |
| 5 | 2023-06-04 | A | 15 | 14 |  |
| 6 | 2023-10-23 | A | 10 | 13,33333 |  |
| 7 | 2024-01-18 | A | 5 | 12,5 | SellId 1 is now out of range |
| 8 | 2024-03-02 | A | 25 | 13,75 | SellId 4 is now out of range |
| 9 | 2024-03-02 | A | 25 | 16 |  |
| 10 | 2024-05-06 | A | 5 | 14,16667 |  |

Let's take "sellId" `3` for example. Even though the "sellTimestamp" (simplified to date) is the same as for "sellId" `2` and `4`, only the "turnover" up to and including "sellId" `3` should be the range for calculating the average.

From "sellId" `7` on "sellId" `1` is out of range given a 365 days interval.

I will check on the window groups clause.

**Reply by UnamedRus** [reply]
> From "sellId" 7 on "sellId" 1 is out of range given a 365 days interval.

After GROUP BY, you can do right join with generated table which contains 1 row per each day-item_id, it will be reasonable small

 > Let's take "sellId" 3 for example. Even though the "sellTimestamp" (simplified to date) is the same as for "sellId" 2 and 4, only the "turnover" up to and including "sellId" 3 should be the range for calculating the average.
 
 
 >  same as for "sellId" 2 and 4, only the "turnover" up to and including "sellId" 3 should be the range for calculating the average.
 
 In that case, set window frame to preceding 365 and  preceding 1
 And collect turnover for current day, in array, just like with sellId, in that way it will allow you in outer query add only previous sellIds from current day

**Reply by UnamedRus** [reply]
BTW, RANGE FRAME is supported https://fiddle.clickhouse.com/3b26953e-8335-4169-846b-03cfe8d8f0dc
